### ISSUE TITLE
COPY/ADD don't support symlinks properly

### ISSUE THREAD

**Issue by johnl** [docker/docker#17730]
Client:
 Version:      1.9.0
 API version:  1.21
 Go version:   go1.4.2
 Git commit:   76d6bc9
 Built:        Tue Nov  3 17:54:53 UTC 2015
 OS/Arch:      linux/amd64

Server:
 Version:      1.9.0
 API version:  1.21
 Go version:   go1.4.2
 Git commit:   76d6bc9
 Built:        Tue Nov  3 17:54:53 UTC 2015
 OS/Arch:      linux/amd64

Containers: 34
Images: 83
Server Version: 1.9.0
Storage Driver: btrfs
 Build Version: Btrfs v4.0
 Library Version: 101
Execution Driver: native-0.2
Logging Driver: json-file
Kernel Version: 4.2.0-16-generic
Operating System: Ubuntu 15.10
CPUs: 4
Total Memory: 7.505 GiB
Name: dogen
ID: G2RV:SXOU:TWY6:WYZA:TRQK:7OM2:W443:Y4P6:4A6P:YET7:U5O5:VOBS
Username: johnl
Registry: https://index.docker.io/v1/
WARNING: No swap limit support

Linux dogen 4.2.0-16-generic #19-Ubuntu SMP Thu Oct 8 15:35:06 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux

Running on a laptop, with btrfs root filesystem.

To reproduce:

```
touch realfile
ln -s realfile symlink
cat > Dockerfile <<EOF
FROM busybox
ADD symlink /dest/
RUN ls -lah /dest
EOF
```

When I build this I get:
```
$ docker build .

Sending build context to Docker daemon 3.072 kB
Step 1 : FROM busybox
 ---> c51f86c28340
Step 2 : ADD symlink /dest/
 ---> 341f3115134c
Removing intermediate container 8cc8e5408cf9
Step 3 : RUN ls /dest/
 ---> Running in 056b6093740c
realfile
 ---> 36845579d4c3
Removing intermediate container 056b6093740c
Successfully built 36845579d4c3
```

As you can see above, the symlink is not copied, and instead realfile is.

I would expect either:

1. the symlink and the realfile get copied in, with the symlink copied as an actual symlink
or
2. a file named symlink with the contents of realfile should be created

or I suppose raising an error.

I can force the behaviour of 2 by specifying the filename in the destination:
```
ADD symlink /dest/symlink
```

**Comment by mikz** [comment]
There is different, but related issue with symlinks: #17756

**Comment by tonistiigi** [comment]
This should be fixed with #17710

**Comment by tonistiigi** [comment]
Closing as #17710 is merged
